### PR TITLE
run_app and CLI can listen on Unix domain sockets

### PR DESCRIPTION
**Pull request by JustinTArthur**
## What do these changes do?
Makes it easy to serve HTTP or HTTPS requests over Unix domain sockets using pre-existing deployment features. Accomplishes this by:
* Adding a `path=` keyword argument to the `run_app` utility function that can be used to supply a file system path for a Unix domain socket.
* Adding an optional `-U` or `--path` command line argument to the CLI web server that can be used to supply a file system path for a Unix domain socket.

## Are there changes in behavior for the user?
Users serving HTTP or HTTPS over TCP/IP should see no change in behavior.

## Needs Review Before Merge
* Is the term _path_ a good one to use in our APIs? It mimics the `path` parameter passed to asyncio's implementation and is small. Is it intuitive enough?
* CPython <3.6 fails to remove a stale unix socket path from the file system before attempting to bind again, causing an error if a stale path exists (see python/asyncio#425). Do we want to work around this for our 3.4 and 3.5 users and delete the stale path before calling `create_unix_server`?
* I had to choose a URI or URL format to use for portraying the HTTP-over-Unix-domain-socket resource. In this version the text for humans that gets spit out by `run_app` uses [the same format as nginx's proxy_pass directive](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_pass), e.g. `http://unix:/tmp/backend.sock:`. Some other possible locator or identifier formats we could use:
   * `http+unix://%2Ftmp%2Fbackend.sock` as used by [requests-unix](https://github.com/msabramo/requests-unixsocket).
  * Place a dummy host as used by [curl](https://curl.haxx.se/)'s `curl` tool: `http://localhost/`
    * It does this partly because its Unix domain socket is supplied separately.
  * `unix:/tmp/backend` as used in nginx's `upstream` `server` directive.
    * It won't portray whether the socket is using SSL or not.

### CODE REVIEW

File: aiohttp/web.py [hunk 1]
```diff
@@ -289,16 +290,30 @@ def run_app(app, *, host='0.0.0.0', port=None,
...
+        srv_creation = loop.create_unix_server(
+            handler, path, ssl=ssl_context, backlog=backlog
+        )
+        # nginx-style URLs for Unix socket paths
+        if isinstance(path, str):
+            urls = ('{}://unix:{}:'.format(scheme, path),)
+        else:
+            urls = ['{}://unix:{}:'.format(scheme, p) for p in path]
+
+    srv = loop.run_until_complete(srv_creation)
```
Comment (fafhrd91) [inline aiohttp/web.py]: i think you should be able to run both unix and tcp servers at the same time

File: docs/deployment.rst [hunk 2]
```diff
@@ -96,23 +96,24 @@ Next we need to configure *aiohttp upstream group*:
 
    http {
      upstream aiohttp {
-       fail_timeout=0 means we always retry an upstream even if it failed
+       # fail_timeout=0 means we always retry an upstream even if it failed
```
Comment (fafhrd91) [inline docs/deployment.rst:99]: may be there should be two sections one for tcp server and for unix socket server

File: aiohttp/web.py (hunk 1 above)
Comment (JustinTArthur) [reply aiohttp/web.py]: Doesn't sound bad. I will have to consider the impact on the parameters (for example, when assuming a host of '0.0.0.0')

File: docs/deployment.rst (hunk 2 above)
Comment (JustinTArthur) [reply docs/deployment.rst:99]: That would be easy enough to do. I can commit a version like this and we'll evaluate if it's too verbose.

Comment (fafhrd91) [reply docs/deployment.rst:99]: sounds good

File: aiohttp/web.py (hunk 1 above)
Comment (fafhrd91) [reply aiohttp/web.py]: les start tcp server only if host is explicitly provide

**Comment by JustinTArthur** [comment]
I have now…
* Added stale socket file cleanup on start for the older event loop implementations that don't do this.
* Added support for passing a bytes-like object for the host(s) or path(s).
* Added tests to validate Linux abstract path support.
  * The URI that gets printed may not be human-usable in this current version.
* Made it so you can supply as many hosts or paths as you want at the same time per @fafhrd91's recommendation..
  * If none are supplied or only a port is supplied, the default behavior of serving on 0.0.0.0 occurs just as has for a long time.
* Documented code snippets for both TCP sockets and Unix domain sockets in the deployment doc per @fafhrd91's recommendation.
* Got code coverage for the web module back up to par.

@fafhrd91 how does it look?

**Comment by fafhrd91** [comment]
Perfect! Thanks!

**Pull request merged**